File: arduino/z_matlab_connect.m
% esp32_blink_control.m
clear;
clc;

% UUID definitions (must match ESP32)
SERVICE_UUID = '4fafc201-1fb5-459e-8fcc-c5c9c331914b';
CHAR_UUID = 'beb5483e-36e1-4688-b7f5-ea07361b26a8';

try
    % Scan for BLE devices
    fprintf('Scanning for ESP32_BLINK...\n');
    b = ble('ESP32_BLINK');
    
    % Connect to the service
    charac = characteristic(b, SERVICE_UUID, CHAR_UUID);
    
    fprintf('Connected to ESP32!\n');
    fprintf('Available commands:\n');
    fprintf('1. Type "START" to begin LED blinking\n');
    fprintf('2. Type "STOP" to stop LED blinking\n');
    fprintf('3. Type "INTERVAL" followed by milliseconds (e.g., "INTERVAL 1000")\n');
    fprintf('4. Type "EXIT" to quit\n\n');
    
    while true
        % Get command from user
        cmd = input('Enter command: ', 's');
        cmd = upper(cmd);  % Convert to uppercase
        
        % Process command
        if startsWith(cmd, 'INTERVAL')
            % Parse interval command
            parts = split(cmd);
            if length(parts) == 2
                interval = str2double(parts{2});
                if ~isnan(interval) && interval > 0
                    write(charac, unicode2native(['INTERVAL:' parts{2}], 'ASCII'));
                    fprintf('Set blink interval to %d ms\n', interval);
                else
                    fprintf('Invalid interval value!\n');
                end
            else
                fprintf('Invalid command format! Use "INTERVAL 1000"\n');
            end
        else
            switch cmd
                case 'START'
                    write(charac, unicode2native('START', 'ASCII'));
                    fprintf('Started LED blinking\n');

                case 'STOP'
                    write(charac, unicode2native('STOP', 'ASCII'));
                    fprintf('Stopped LED blinking\n');

                case 'EXIT'
                    fprintf('Disconnecting...\n');
                    clear b;
                    break;

                otherwise
                    fprintf('Invalid command!\n');
            end
        end
    end
    
catch ME
    fprintf('Error: %s\n', ME.message);
    if exist('b', 'var')
        clear b;
    end
end

fprintf('Program ended\n');

File: embedded/scripts_python/z_matlab_connect.py
import asyncio
from bleak import BleakClient, BleakScanner
import sys

# UUID definitions (must match ESP32)
SERVICE_UUID = "4fafc201-1fb5-459e-8fcc-c5c9c331914b"
CHAR_UUID = "beb5483e-36e1-4688-b7f5-ea07361b26a8"

async def main():
    try:
        # Scan for BLE devices
        print("Scanning for ESP32_BLINK...")
        device = await BleakScanner.find_device_by_name("ESP32_BLINK")
        
        if device is None:
            print("Could not find ESP32_BLINK device!")
            return

        async with BleakClient(device) as client:
            print("Connected to ESP32!")
            print("Available commands:")
            print('1. Type "START" to begin LED blinking')
            print('2. Type "STOP" to stop LED blinking')
            print('3. Type "INTERVAL" followed by milliseconds (e.g., "INTERVAL 1000")')
            print('4. Type "EXIT" to quit\n')

            while True:
                # Get command from user
                cmd = input("Enter command: ").upper()

                # Process command
                if cmd.startswith("INTERVAL"):
                    # Parse interval command
                    parts = cmd.split()
                    if len(parts) == 2:
                        try:
                            interval = int(parts[1])
                            if interval > 0:
                                await client.write_gatt_char(CHAR_UUID, f"INTERVAL:{interval}".encode('ascii'))
                                print(f"Set blink interval to {interval} ms")
                            else:
                                print("Invalid interval value!")
                        except ValueError:
                            print("Invalid interval value!")
                    else:
                        print('Invalid command format! Use "INTERVAL 1000"')
                
                else:
                    if cmd == "START":
                        await client.write_gatt_char(CHAR_UUID, b"START")
                        print("Started LED blinking")
                    
                    elif cmd == "STOP":
                        await client.write_gatt_char(CHAR_UUID, b"STOP")
                        print("Stopped LED blinking")
                    
                    elif cmd == "EXIT":
                        print("Disconnecting...")
                        break
                    
                    else:
                        print("Invalid command!")

    except Exception as e:
        print(f"Error: {str(e)}")

    print("Program ended")

if __name__ == "__main__":
    # Run the async main function
    asyncio.run(main())